### DISCUSSION TITLE
Parser Documentation

### DISCUSSION THREAD

**Discussion by wolframm** [petitparser/dart-petitparser#165] [General]
Is it possible to generate human-readable syntax diagrams or some human-readable syntax definition as documentation for a parser built with PetitParser?

**Comment by renggli** [comment]
This is certainly possible, given the the [reflective nature](https://pub.dev/documentation/petitparser/latest/reflection/reflection-library.html) of parsers. The Smalltalk version of PetitParser had a couple of visualizations and even a visual IDE, check _3.5 Tool Support_ in [this paper](https://scg.unibe.ch/archive/papers/Reng10cDynamicGrammars.pdf).

I am not aware that this has been done with PetitParser for Dart yet, but creating a simple visualization for example using [Graphviz](https://graphviz.org/) should be easy: use [allParser](https://pub.dev/documentation/petitparser/latest/reflection/allParser.html) to fetch the nodes of the graph, and query [children](https://pub.dev/documentation/petitparser/latest/core/Parser/children.html) to get all outgoing edges.
